src/util/receipt.ts

```diff
@@ -0,0 +1,71 @@
+import { Reader, Writer } from 'oer-utils'
+import {
+  LongValue,
+  longFromValue
+} from './long'
+import * as Long from 'long'
+import { generateReceiptHMAC } from '../crypto'
+
+const RECEIPT_VERSION = 1
+export { RECEIPT_VERSION }
+
+export interface ReceiptOpts {
+  nonce: Buffer
+  streamId: LongValue
+  totalReceived: LongValue
+  secret: Buffer
+}
+
+export interface Receipt {
+  version: number
+  nonce: Buffer
+  streamId: string
+  totalReceived: Long
+}
+
+export function createReceipt (opts: ReceiptOpts): Buffer {
+  if (opts.nonce.length !== 16) {
+    throw new Error('receipt nonce must be 16 bytes')
+  }
+  if (opts.secret.length !== 32) {
+    throw new Error('receipt secret must be 32 bytes')
+  }
+  const receipt = new Writer(58)
+  receipt.writeUInt8(RECEIPT_VERSION)
+  receipt.writeOctetString(opts.nonce, 16)
+  receipt.writeUInt8(opts.streamId)
+  receipt.writeUInt64(longFromValue(opts.totalReceived, true))
+  receipt.writeOctetString(generateReceiptHMAC(opts.secret, receipt.getBuffer()), 32)
+  return receipt.getBuffer()
+}
+
+export function decodeReceipt (receipt: Buffer): Receipt {
+  if (receipt.length !== 58) {
+    throw new Error('receipt must be 58 bytes')
+  }
+  const reader = Reader.from(receipt)
+  const version = reader.readUInt8Number()
+  const nonce = reader.readOctetString(16)
+  const streamId = reader.readUInt8()
+  const totalReceived = reader.readUInt64Long()
+  return {
+    version,
+    nonce,
+    streamId,
+    totalReceived
+  }
+}
+
+export function verifyReceipt (receipt: Buffer, secret: Buffer): Boolean {
```

Comment:
we might want something like the JWT library's decode, which validates and also returns the deserialized fields

Reply:
Updated `verifyReceipt` to return decoded fields like ([JWT's verify](https://github.com/auth0/node-jsonwebtoken#jwtverifytoken-secretorpublickey-options-callback))
https://github.com/interledgerjs/ilp-protocol-stream/pull/133/commits/7468b13d2b4d4e4044004c727509d1131e5d70d1